include/LightGBM/category_encoding_provider.hpp

```diff
@@ -0,0 +1,771 @@
+  * \brief Extends the settings per feature to include the category encoding features
+  * \param config Config from boosting object
+  */
+  void ExtendPerFeatureSetting(Config* config) const;
+
+ private:
+  void SetConfig(const Config* config);
+
+  explicit CategoryEncodingProvider(const std::string model_string);
+
+  CategoryEncodingProvider(const char* str, size_t* used_len);
+
+  explicit CategoryEncodingProvider(Config* config);
+
+  CategoryEncodingProvider(Config* config,
+    const std::vector<std::function<std::vector<double>(int row_idx)>>& get_row_fun,
+    const std::function<label_t(int row_idx)>& get_label_fun, const int32_t nmat,
+    const int32_t* nrow, const int32_t ncol);
+
+  CategoryEncodingProvider(Config* config,
+    const std::function<std::vector<std::pair<int, double>>(int row_idx)>& get_row_fun,
+    const std::function<label_t(int row_idx)>& get_label_fun,
+    const int64_t nrow, const int64_t ncol);
+
+  CategoryEncodingProvider(Config* config,
+    const std::vector<std::unique_ptr<CSC_RowIterator>>& csc_iters,
+    const std::function<label_t(int row_idx)>& get_label_fun,
+    const int64_t nrow, const int64_t ncol);
+
+  template <bool ACCUMULATE_FROM_FILE>
+  void ProcessOneLineInner(const std::vector<std::pair<int, double>>& one_line,
+    double label, int line_idx, std::vector<bool>* is_feature_processed_ptr,
+    std::unordered_map<int, std::vector<std::unordered_map<int, int>>>* count_info_ptr,
+    std::unordered_map<int, std::vector<std::unordered_map<int, double>>>* label_info_ptr,
+    std::vector<double>* label_sum_ptr, std::vector<int>* num_data_ptr, const int fold_id);
+
+  // sync up encoding values by gathering statistics from all machines in distributed scenario
+  void SyncEncodingStat(std::vector<std::unordered_map<int, double>>* fold_label_sum_ptr,
+    std::vector<std::unordered_map<int, int>>* fold_total_count_ptr, const int num_machines) const;
+
+  // sync up statistics to calculate the encoding prior by gathering statistics from all machines in distributed scenario
+  void SyncEncodingPrior(const double label_sum, const int local_num_data, double* all_label_sum_ptr,
+    int* all_num_data_ptr, int num_machines) const;
+
+  int ParseMetaInfo(const char* filename, Config* config);
+
+  void ExpandNumFeatureWhileAccumulate(const int new_largest_fid);
+
+  inline void AddCountAndLabel(std::unordered_map<int, int>* count_map,
+    std::unordered_map<int, double>* label_map,
+    const int cat_value, const int count_value, const label_t label_value) {
+    if (count_map->count(cat_value) == 0) {
+      count_map->operator[](cat_value) = count_value;
+      label_map->operator[](cat_value) = static_cast<double>(label_value);
+    } else {
+      count_map->operator[](cat_value) += count_value;
+      label_map->operator[](cat_value) += static_cast<double>(label_value);
+    }
+  }
+
+  // parameter configuration
+  Config config_;
+
+  // size of training data
+  data_size_t num_data_;
```

Comment:
for below properties we'd better have only 1 unified pair serialization & deserialization function?